client/servicediscovery/pd_service_discovery.go

```diff
@@ -509,8 +509,8 @@ func (c *pdServiceDiscovery) Init() error {
 		}
 	}
 
-	if err := c.checkServiceModeChanged(); err != nil {
-		log.Warn("[pd] failed to check service mode and will check later", zap.Error(err))
+	if err := c.initRetry(c.checkServiceModeChanged); err != nil {
+		return err
```

Comment:
One more thing, we need to handle cancel properly.

Reply:
which `cancel`?

Reply:
See line 499 and line 495, but I find the caller will also call cancel.

Reply:
Added https://github.com/tikv/pd/pull/8963/commits/81261bdc9ce9334acf2d3358c5b6aa8b65112d0b